core/src/database/header_chain_prover.rs

```diff
@@ -35,76 +35,188 @@ impl Database {
+    ///
+    /// # Returns
+    ///
+    /// Returns `None` if either no proved blocks are exists or blockchain tip
+    /// is already proven.
+    ///
+    /// - [`BlockHash`] - Hash of the block
+    /// - [`Header`] - Header of the block
+    /// - [`u64`] - Height of the block
+    /// - [`Receipt`] - Previous block's proof
+    pub async fn get_next_unproven_block(
+        &self,
+        tx: Option<DatabaseTransaction<'_, '_>>,
+    ) -> Result<Option<(BlockHash, Header, u64, Receipt)>, BridgeError> {
         let query = sqlx::query_as(
-            "SELECT block_header FROM header_chain_proofs WHERE height = $1 AND block_hash = $2;",
-        )
-        .bind(block_height as i64)
-        .bind(BlockHashDB(block_hash));
+            "SELECT h1.block_hash,
+                    h1.block_header,
+                    h1.height,
+                    h2.proof
+                FROM header_chain_proofs h1
+                JOIN header_chain_proofs h2 ON h1.prev_block_hash = h2.block_hash
+                WHERE h2.proof IS NOT NULL AND h1.proof IS NULL
+                ORDER BY h1.height DESC
+                LIMIT 1;",
+        );
 
-        let result: (Option<BlockHeaderDB>,) =
-            execute_query_with_tx!(self.connection, tx, query, fetch_one)?;
+        let result: Option<(BlockHashDB, BlockHeaderDB, i64, Vec<u8>)> =
+            execute_query_with_tx!(self.connection, tx, query, fetch_optional)?;
+
+        let result = match result {
+            Some(result) => {
+                let receipt: Receipt =
+                    borsh::from_slice(&result.3).wrap_err(BridgeError::BorshError)?;
+                let height = result.2.try_into().wrap_err("Can't convert i64 to u64")?;
+                Some((result.0 .0, result.1 .0, height, receipt))
+            }
+            None => None,
+        };
+
+        Ok(result)
+    }
 
-        match result {
-            (Some(block_header),) => Ok(Some(block_header.0)),
-            (None,) => Ok(None),
+    /// Gets the newest n number of block's info that their previous block has
+    /// proven before. These blocks will be the candidate blocks for the prover.
+    ///
+    /// # Returns
+    ///
+    /// Returns `None` if either no proved blocks are exists or blockchain tip
+    /// is already proven.
+    ///
+    /// - [`BlockHash`] - Hash of last block in the batch
+    /// - [`Header`] - Headers of the blocks
+    /// - [`u64`] - Height of the last block in the batch
+    /// - [`Receipt`] - Previous block's proof
+    pub async fn get_next_n_non_proven_block(
+        &self,
+        count: u32,
+    ) -> Result<Option<(Vec<(BlockHash, Header, u64)>, Receipt)>, BridgeError> {
+        let next_non_proven_block = self.get_next_unproven_block(None).await?;
+        let next_non_proven_block = match next_non_proven_block {
+            Some(next_non_proven_block) => next_non_proven_block,
+            None => return Ok(None),
+        };
```

Comment:
let Some(next_non_proven_block) = self.get_next_unproven_block(None).await? else {return Ok(None)};

is clearer